### PR TITLE
Enable override checking for type methods and methods with param/type return types

### LINKED ISSUE #13154: how are type methods inherited?
This language-design issue raises these related questions:

(A) Should a type method defined on a parent class be invocable on its child class?

(B) Should a child class be allowed to override a type method on its parent class?

(C) If yes on A and B, should a type method in the child class be required or allowed or disallowed to have an `override` annotation?

(D) If yes on A and B, what type should be `this` inside a type method defined on a parent class and invoked on a child class?

C++, Swift, and present-day Chapel answer YES to (A) and (B).

Swift requires `override`, which I am inferring from [this page](https://docs.swift.org/swift-book/LanguageGuide/Inheritance.html), where it says:

> A subclass can provide its own custom implementation of [...] type method [...] that it would otherwise inherit from a superclass. This is known as overriding. [...] any overrides without the override keyword are diagnosed as an error [...].

Tests: there are tests/futures in `test/functions/typeMethods`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR enables override checking for type methods, methods that return `param`, and methods that return `type`.

These methods behave like normal class methods with respect to the `override` keyword, but they are not subject to virtual dispatch. Type methods and methods with param/type return types are resolved statically. Even if a subclass override of a param/type return method has been defined, the actual routine to be called depends on the static type of the receiver.

Resolves #13154.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

File: compiler/resolution/virtualDispatch.cpp [hunk 1]
```diff
@@ -527,19 +528,21 @@ static bool isSubType(Type* sub, Type* super) {
...
-               !fn->hasFlag(FLAG_NO_PARENS) &&
-                fn->retTag != RET_PARAM &&
-                fn->retTag != RET_TYPE;
+  if (AggregateType* at = getReceiverClassType(fn)) {
+    INT_ASSERT(at->isClass()); 
+    return fn->name != astrInit &&
+           !fn->hasFlag(FLAG_WRAPPER) &&
+           !fn->hasFlag(FLAG_NO_PARENS) &&
+            fn->retTag != RET_PARAM &&
+            fn->retTag != RET_TYPE;
```
Comment (mppf) [inline compiler/resolution/virtualDispatch.cpp]: I'm not so confident that you want to have RET_PARAM and RET_TYPE apply to your type methods. Try making some tests with e.g. `proc type method() type { }` and `proc type method() param { }`

Comment (dlongnecke-cray) [reply compiler/resolution/virtualDispatch.cpp]: Thanks, I'll add a few more tests.

**Comment by dlongnecke-cray** [comment]
Just a heads up, after enabling override checking for type/param return types, I have over 1300 test failures to fix. Most are hopefully related to arrays, we'll see what things look like after I whittle them down.

**Comment by mppf** [comment]
@dlongnecke-cray - in that case, I recommend merging this PR without it and then following up in a separate PR.

**Comment by dlongnecke-cray** [comment]
Whoops, a bit too late for that, as I've already made some progress on the weekend and reoriented the PR message.

I agree that's probably the correct thing to do in the future, though.

**Comment by dlongnecke-cray** [comment]
@bradcray Pinging you after discussion.

**Comment by bradcray** [comment]
Thanks David.  As you said, this is all looking like a no-brainer to me.  I'll be curious if any of the remaining test failures require changes to the tests themselves, and if so, whether any of those give us pause (but am not expecting it).

**Comment by dlongnecke-cray** [comment]
Nope, so far it's just been some 30~ odd cases of inserting the override keyword for methods that return param/type.

I went ahead and added a new check `isVirtualizableMethod` to the virtual dispatch code that preserves the old semantics for virtual dispatch, so I'm not particularly worried about something unexpected popping up.

**Comment by dlongnecke-cray** [comment]
Hi @mppf, could you do one last quick check to make sure things look OK w.r.t. to the `override` keyword additions before I merge?

**Pull request merged**